File: Paper2Codes-WebUI/lib/api/analytics.ts
/**
 * Analytics API client
 */

import { apiClient } from './client';
import { API_BASE_URL } from '@/lib/config';
import { ApiClientError } from './errors';
import type { ApiResponse } from './types';

export interface AnalyticsOverview {
  total_papers: number;
  total_repositories: number;
  total_tasks: number;
  active_tasks: number;
  completed_tasks: number;
  failed_tasks: number;
  total_modules: number;
  total_verifications: number;
  average_generation_time_seconds: number;
  average_verification_time_seconds: number;
  success_rate: number;
}

export interface PerformanceMetrics {
  api_response_times: ResponseTimeMetrics;
  database_query_times: ResponseTimeMetrics;
  llm_request_times: ResponseTimeMetrics;
  task_execution_times: TaskExecutionMetrics;
  cache_hit_rate: number;
  throughput: ThroughputMetrics;
}

export interface ResponseTimeMetrics {
  p50_ms: number;
  p95_ms: number;
  p99_ms: number;
  average_ms: number;
  min_ms: number;
  max_ms: number;
}

export interface TaskExecutionMetrics {
  planning_avg_seconds: number;
  analysis_avg_seconds: number;
  coding_avg_seconds: number;
  verification_avg_seconds: number;
  total_avg_seconds: number;
}

export interface ThroughputMetrics {
  requests_per_second: number;
  tasks_per_hour: number;
  papers_processed_per_day: number;
}

export interface UsageStatistics {
  period: string;
  total_requests: number;
  total_tasks_created: number;
  total_papers_uploaded: number;
  total_repositories_generated: number;
  llm_requests: number;
  llm_tokens_used: number;
  storage_used_bytes: number;
  daily_breakdown: Array<{
    date: string;
    requests: number;
    tasks: number;
    papers: number;
    repositories: number;
  }>;
}

export interface AgentPerformanceMetrics {
  planning_agent: AgentMetrics;
  analysis_agent: AgentMetrics;
  coding_agent: AgentMetrics;
  verification_agent: AgentMetrics;
}

export interface AgentMetrics {
  total_executions: number;
  successful_executions: number;
  failed_executions: number;
  average_execution_time_seconds: number;
  average_tokens_used: number;
  average_cost_usd: number;
  success_rate: number;
}

/**
 * Get analytics overview
 */
export async function getAnalyticsOverview(): Promise<
  ApiResponse<AnalyticsOverview>
> {
  return apiClient.getAnalyticsOverview();
}

/**
 * Get performance metrics
 */
export async function getPerformanceMetrics(
  range?: string
): Promise<ApiResponse<PerformanceMetrics>> {
  return apiClient.getPerformanceMetrics(range);
}

/**
 * Get usage statistics
 */
export async function getUsageStatistics(
  period?: string
): Promise<ApiResponse<UsageStatistics>> {
  return apiClient.getUsageStatistics(period);
}

/**
 * Get agent performance metrics
 */
export async function getAgentPerformance(
  range?: string
): Promise<ApiResponse<AgentPerformanceMetrics>> {
  return apiClient.getAgentPerformance(range);
}

/**
 * Get time series data for charts
 */
export async function getTimeSeriesData(
  metric: 'tasks' | 'papers' | 'requests' | 'llm_usage',
  range: '24h' | '7d' | '30d' | '90d' = '7d'
): Promise<
  ApiResponse<
    Array<{
      timestamp: string;
      value: number;
      metadata?: Record<string, unknown>;
    }>
  >
> {
  return apiClient.get('/api/analytics/timeseries', {
    params: { metric, range },
  });
}

/**
 * Get domain distribution
 */
export async function getDomainDistribution(): Promise<
  ApiResponse<
    Array<{
      domain: string;
      count: number;
      percentage: number;
    }>
  >
> {
  return apiClient.get('/api/analytics/domains');
}

/**
 * Get LLM usage breakdown
 */
export async function getLLMUsageBreakdown(): Promise<
  ApiResponse<
    Array<{
      model: string;
      requests: number;
      tokens: number;
      cost_usd: number;
      percentage: number;
    }>
  >
> {
  return apiClient.get('/api/analytics/llm-usage');
}

/**
 * Export analytics report
 */
export async function exportAnalyticsReport(
  format: 'csv' | 'json' | 'pdf',
  range?: string
): Promise<Blob> {
  if (typeof window === 'undefined') {
    throw new ApiClientError(
      'UNSUPPORTED_ENVIRONMENT',
      'Analytics export is only supported in the browser',
    );
  }

  const url = new URL('/api/analytics/export', API_BASE_URL);
  url.searchParams.append('format', format);
  if (range) {
    url.searchParams.append('range', range);
  }

  const token = localStorage.getItem('auth_token');
  const headers: Record<string, string> = {};
  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  const response = await fetch(url.toString(), {
    headers,
    credentials: 'include',
  });
  if (!response.ok) {
    const errorText = await response.text().catch(() => '');
    throw new ApiClientError(
      'EXPORT_FAILED',
      'Failed to export analytics report',
      response.status,
      { endpoint: url.pathname, responseBody: errorText || undefined },
    );
  }

  return response.blob();
}

File: Paper2Codes-Core/src/api/handlers/analytics.rs
//! Analytics and metrics handlers for Phase 5
//!
//! Provides analytics endpoints for overview, performance, usage, and agent statistics

use axum::{
    extract::{Extension, Query},
    response::Json,
};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tracing::info;

use crate::api::state::AppState;
use crate::api::types::responses::ApiResponse;
use crate::error::Result;

/// Analytics overview response
#[derive(Debug, Serialize, Deserialize)]
pub struct AnalyticsOverview {
    pub total_papers: usize,
    pub total_repositories: usize,
    pub total_tasks: usize,
    pub active_tasks: usize,
    pub completed_tasks: usize,
    pub failed_tasks: usize,
    pub total_modules: usize,
    pub total_verifications: usize,
    pub average_generation_time_seconds: f64,
    pub average_verification_time_seconds: f64,
    pub success_rate: f64,
}

/// Performance metrics response
#[derive(Debug, Serialize, Deserialize)]
pub struct PerformanceMetrics {
    pub api_response_times: ResponseTimeMetrics,
    pub database_query_times: ResponseTimeMetrics,
    pub llm_request_times: ResponseTimeMetrics,
    pub task_execution_times: TaskExecutionMetrics,
    pub cache_hit_rate: f64,
    pub throughput: ThroughputMetrics,
}

/// Response time metrics
#[derive(Debug, Serialize, Deserialize)]
pub struct ResponseTimeMetrics {
    pub p50_ms: f64,
    pub p95_ms: f64,
    pub p99_ms: f64,
    pub average_ms: f64,
    pub min_ms: f64,
    pub max_ms: f64,
}

/// Task execution metrics
#[derive(Debug, Serialize, Deserialize)]
pub struct TaskExecutionMetrics {
    pub planning_avg_seconds: f64,
    pub analysis_avg_seconds: f64,
    pub coding_avg_seconds: f64,
    pub verification_avg_seconds: f64,
    pub total_avg_seconds: f64,
}

/// Throughput metrics
#[derive(Debug, Serialize, Deserialize)]
pub struct ThroughputMetrics {
    pub requests_per_second: f64,
    pub tasks_per_hour: f64,
    pub papers_processed_per_day: f64,
}

/// Usage statistics response
#[derive(Debug, Serialize, Deserialize)]
pub struct UsageStatistics {
    pub period: String, // "day", "week", "month", "year"
    pub total_requests: usize,
    pub total_tasks_created: usize,
    pub total_papers_uploaded: usize,
    pub total_repositories_generated: usize,
    pub llm_requests: usize,
    pub llm_tokens_used: usize,
    pub storage_used_bytes: usize,
    pub daily_breakdown: Vec<DailyUsage>,
}

/// Daily usage breakdown
#[derive(Debug, Serialize, Deserialize)]
pub struct DailyUsage {
    pub date: String,
    pub requests: usize,
    pub tasks: usize,
    pub papers: usize,
    pub repositories: usize,
}

/// Agent performance metrics
#[derive(Debug, Serialize, Deserialize)]
pub struct AgentPerformanceMetrics {
    pub planning_agent: AgentMetrics,
    pub analysis_agent: AgentMetrics,
    pub coding_agent: AgentMetrics,
    pub verification_agent: AgentMetrics,
}

/// Agent-specific metrics
#[derive(Debug, Serialize, Deserialize)]
pub struct AgentMetrics {
    pub total_executions: usize,
    pub successful_executions: usize,
    pub failed_executions: usize,
    pub average_execution_time_seconds: f64,
    pub average_tokens_used: usize,
    pub average_cost_usd: f64,
    pub success_rate: f64,
}

/// Get analytics overview
///
/// Endpoint: GET /api/analytics/overview
pub async fn get_analytics_overview(
    Extension(_state): Extension<Arc<AppState>>,
) -> Result<Json<ApiResponse<AnalyticsOverview>>> {
    info!("Analytics overview request");

    // TODO: Implement actual analytics aggregation from storage
    // This should query SurrealDB for:
    // - Count of papers, repositories, tasks, modules
    // - Task status breakdown
    // - Average times
    // - Success rates

    let overview = AnalyticsOverview {
        total_papers: 100,
        total_repositories: 50,
        total_tasks: 500,
        active_tasks: 10,
        completed_tasks: 450,
        failed_tasks: 40,
        total_modules: 1000,
        total_verifications: 400,
        average_generation_time_seconds: 120.0,
        average_verification_time_seconds: 30.0,
        success_rate: 0.9,
    };

    Ok(Json(ApiResponse::new(overview)))
}

/// Get performance metrics
///
/// Endpoint: GET /api/analytics/performance
pub async fn get_performance_metrics(
    Query(params): Query<HashMap<String, String>>,
    Extension(_state): Extension<Arc<AppState>>,
) -> Result<Json<ApiResponse<PerformanceMetrics>>> {
    let time_range = params
        .get("range")
        .cloned()
        .unwrap_or_else(|| "24h".to_string());
    info!("Performance metrics request: range={}", time_range);

    // TODO: Implement actual performance metrics from monitoring/logging
    // This should aggregate:
    // - Response times (p50, p95, p99)
    // - Database query performance
    // - LLM request performance
    // - Task execution times
    // - Cache hit rates
    // - Throughput metrics

    let metrics = PerformanceMetrics {
        api_response_times: ResponseTimeMetrics {
            p50_ms: 50.0,
            p95_ms: 150.0,
            p99_ms: 300.0,
            average_ms: 75.0,
            min_ms: 10.0,
            max_ms: 500.0,
        },
        database_query_times: ResponseTimeMetrics {
            p50_ms: 10.0,
            p95_ms: 30.0,
            p99_ms: 50.0,
            average_ms: 15.0,
            min_ms: 1.0,
            max_ms: 100.0,
        },
        llm_request_times: ResponseTimeMetrics {
            p50_ms: 2000.0,
            p95_ms: 5000.0,
            p99_ms: 8000.0,
            average_ms: 2500.0,
            min_ms: 500.0,
            max_ms: 10000.0,
        },
        task_execution_times: TaskExecutionMetrics {
            planning_avg_seconds: 30.0,
            analysis_avg_seconds: 45.0,
            coding_avg_seconds: 60.0,
            verification_avg_seconds: 25.0,
            total_avg_seconds: 160.0,
        },
        cache_hit_rate: 0.75,
        throughput: ThroughputMetrics {
            requests_per_second: 10.0,
            tasks_per_hour: 50.0,
            papers_processed_per_day: 10.0,
        },
    };

    Ok(Json(ApiResponse::new(metrics)))
}

/// Get usage statistics
///
/// Endpoint: GET /api/analytics/usage?period=...
pub async fn get_usage_statistics(
    Query(params): Query<HashMap<String, String>>,
    Extension(_state): Extension<Arc<AppState>>,
) -> Result<Json<ApiResponse<UsageStatistics>>> {
    let period = params
        .get("period")
        .cloned()
        .unwrap_or_else(|| "week".to_string());
    info!("Usage statistics request: period={}", period);

    // TODO: Implement actual usage statistics from storage
    // This should aggregate:
    // - Total requests, tasks, papers, repositories
    // - LLM usage (requests, tokens)
    // - Storage usage
    // - Daily breakdown

    let statistics = UsageStatistics {
        period: period.clone(),
        total_requests: 10000,
        total_tasks_created: 500,
        total_papers_uploaded: 100,
        total_repositories_generated: 50,
        llm_requests: 5000,
        llm_tokens_used: 10000000,
        storage_used_bytes: 1073741824, // 1GB
        daily_breakdown: vec![DailyUsage {
            date: "2024-01-01".to_string(),
            requests: 1000,
            tasks: 50,
            papers: 10,
            repositories: 5,
        }],
    };

    Ok(Json(ApiResponse::new(statistics)))
}

/// Get agent performance metrics
///
/// Endpoint: GET /api/analytics/agents
pub async fn get_agent_performance(
    Query(params): Query<HashMap<String, String>>,
    Extension(_state): Extension<Arc<AppState>>,
) -> Result<Json<ApiResponse<AgentPerformanceMetrics>>> {
    let time_range = params
        .get("range")
        .cloned()
        .unwrap_or_else(|| "24h".to_string());
    info!("Agent performance metrics request: range={}", time_range);

    // TODO: Implement actual agent performance metrics
    // This should aggregate from coordinator/agent metrics:
    // - Execution counts
    // - Success/failure rates
    // - Average execution times
    // - Token usage
    // - Cost calculations

    let agent_metrics = AgentPerformanceMetrics {
        planning_agent: AgentMetrics {
            total_executions: 100,
            successful_executions: 95,
            failed_executions: 5,
            average_execution_time_seconds: 30.0,
            average_tokens_used: 5000,
            average_cost_usd: 0.10,
            success_rate: 0.95,
        },
        analysis_agent: AgentMetrics {
            total_executions: 200,
            successful_executions: 190,
            failed_executions: 10,
            average_execution_time_seconds: 45.0,
            average_tokens_used: 8000,
            average_cost_usd: 0.15,
            success_rate: 0.95,
        },
        coding_agent: AgentMetrics {
            total_executions: 300,
            successful_executions: 270,
            failed_executions: 30,
            average_execution_time_seconds: 60.0,
            average_tokens_used: 10000,
            average_cost_usd: 0.20,
            success_rate: 0.90,
        },
        verification_agent: AgentMetrics {
            total_executions: 150,
            successful_executions: 145,
            failed_executions: 5,
            average_execution_time_seconds: 25.0,
            average_tokens_used: 3000,
            average_cost_usd: 0.05,
            success_rate: 0.97,
        },
    };

    Ok(Json(ApiResponse::new(agent_metrics)))
}

/// Get domain distribution
///
/// Endpoint: GET /api/analytics/domains
pub async fn get_domain_distribution(
    Extension(_state): Extension<Arc<AppState>>,
) -> Result<Json<ApiResponse<Vec<DomainStats>>>> {
    info!("Domain distribution request");

    // TODO: Implement actual domain distribution from storage
    let domains = vec![
        DomainStats {
            domain: "Deep Learning".to_string(),
            count: 35,
            percentage: 35.0,
        },
        DomainStats {
            domain: "NLP".to_string(),
            count: 25,
            percentage: 25.0,
        },
        DomainStats {
            domain: "Computer Vision".to_string(),
            count: 20,
            percentage: 20.0,
        },
        DomainStats {
            domain: "Other".to_string(),
            count: 20,
            percentage: 20.0,
        },
    ];

    Ok(Json(ApiResponse::new(domains)))
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DomainStats {
    pub domain: String,
    pub count: usize,
    pub percentage: f64,
}

/// Get LLM usage breakdown
///
/// Endpoint: GET /api/analytics/llm-usage
pub async fn get_llm_usage_breakdown(
    Extension(_state): Extension<Arc<AppState>>,
) -> Result<Json<ApiResponse<Vec<LLMUsageStats>>>> {
    info!("LLM usage breakdown request");

    // TODO: Implement actual LLM usage from storage
    let usage = vec![
        LLMUsageStats {
            model: "GPT-4 Turbo".to_string(),
            requests: 5000,
            tokens: 124500,
            cost_usd: 24.90,
            percentage: 45.0,
        },
        LLMUsageStats {
            model: "Claude 3 Opus".to_string(),
            requests: 3000,
            tokens: 89300,
            cost_usd: 17.86,
            percentage: 30.0,
        },
        LLMUsageStats {
            model: "Grok-2".to_string(),
            requests: 1500,
            tokens: 42100,
            cost_usd: 8.42,
            percentage: 15.0,
        },
        LLMUsageStats {
            model: "Other".to_string(),
            requests: 1000,
            tokens: 28500,
            cost_usd: 5.70,
            percentage: 10.0,
        },
    ];

    Ok(Json(ApiResponse::new(usage)))
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LLMUsageStats {
    pub model: String,
    pub requests: usize,
    pub tokens: usize,
    pub cost_usd: f64,
    pub percentage: f64,
}
